code/json.sls

```diff
@@ -30,172 +30,233 @@
-	  (let* ((entry (car vec))
-		 (k (car entry))
-		 (v (cdr entry)))
-	    (cond
-	      ((symbol? k) (write (symbol->string k) p))
-	      ((string? k) (write k p)) ;; for convenience
-	      (else (error "Invalid JSON table key in json-write" k)))
-	    (display ": " p)
-	    (write-any v p)))
-	(display "}" p))
+        (display "{" p)
+        (when pretty?
+          (display "\n" p)
+          (set! indent-level (+ indent-level 1)))
+        (do ((need-comma #f #t)
+             (vec vec (cdr vec)))
+            ((null? vec))
+          (if need-comma
+              (begin (display "," p)
+                     (if pretty?
+                         (display "\n" p)
+                         (display " " p)))
+              (set! need-comma #t))
+          (let* ((entry (car vec))
+                 (k (car entry))
+                 (v (cdr entry)))
+            (when pretty? (display-level p))
+            (display "\"" p)
+            (cond
+             ((symbol? k) (display (symbol->string k) p))
+             ((string? k) (display k p))
+             (else (error "Invalid JSON table key in json-write" k)))
+            (display "\": " p)
+            (write-any v p)))
+        (when pretty?
+          (set! indent-level (- indent-level 1))
+          (display "\n" p)
+          (display-level p))
+        (display "}" p))
 
       (define (write-array a p)
-	(display "[" p)
-	(let ((need-comma #f))
-	  (for-each (lambda (v)
-		      (if need-comma
-			  (display ", " p)
-			  (set! need-comma #t))
-		      (write-any v p))
-		    a))
-	(display "]" p))
+        (display "[" p)
+        (when pretty?
+          (display "\n" p)
+          (set! indent-level (+ indent-level 1)))
+        (let ((need-comma #f))
+          (for-each (lambda (v)
+                      (if need-comma
+                          (if pretty?
+                              (display ",\n" p)
+                              (display "," p))
+                          (set! need-comma #t))
+                      (when pretty? (display-level p))
+                      (write-any v p))
+                    a))
+        (when pretty?
+          (set! indent-level (- indent-level 1))
+          (display "\n" p)
+          (display-level p))
+        (display "]" p))
 
       (define (write-any x p)
-	(cond
-	  ((or (string? x)
-	       (number? x)) (write x p))
-	  ((boolean? x) (display (if x "true" "false") p))
-	  ((symbol? x) (write (if (eq? x 'null) 'null (symbol->string x))
-			      p)) ;; for convenience
-	  ((null? x) (display "null" p))
-	  ((and (list? x)
-		(pair? (car x))
-		(not (pair? (caar x))))
-	   (write-ht x p))
-	  ((list? x) (write-array x p))
-	  (else (error "Invalid JSON object in json-write" x))))
+        (cond
+         ((or (string? x)
+              (number? x)) (write x p))
+         ((boolean? x) (display (if x "true" "false") p))
+         ((symbol? x)
+          (write (if (eq? x 'null) 'null (symbol->string x)) p))
```

Comment:
I think `null` is represented as `'()`, so this check might be unnecessary. 
```
> (with-input-from-string "null" json-read)
()
```

Reply:
I think you are right.

This was already in the lib.

I am interested in refactoring this library (and adding better comments to the original author's work), but I think that a full refactor and documentation of his library might be better in a new PR for the future.

Is that fine with you if this change waits until then?  Without a test sweet designed to prove that refactoring doesn't break anything I am concerned there could be edge cases we break by refactoring his code further than the current latch goes.

Reply:
Fair enough, yeah, sounds fine to keep!